### PR TITLE
hashjoin indexing

### PR DESCRIPTION
**Pull request by JCOR11599**
When we have plan of the following format:
```
InSubquery
    ...
    CrossJoin
        Left: Table
        Right: SubqueryAlias
            OuterScopeVisibility: true
            ...
            HashJoin
                HashLookup
                    source: ...
                    target: TableAlias
...
```

The indexes we assign to GetFields during analysis don't align with the indexes of the actual columns in each row during execution time. This is a result of StripNode, PrependNode, and the nested Joins with SubqueryAlias. 

This error wasn't caught sooner as the incorrect indexes are too low, so they never threw `IndexOutOfBounds` errors and just returned potentially incorrect results instead.

The fix was to correct these indexes at analysis time.
Firstly, `SubqueryAlias` nodes with `OuterScopeVisibilty = true` inside joins need to see the left sibling node (in addition to the parent nodes). So `Scope` was modified to include some new fields, specifically for sibling nodes. Additionally, the file `finalizeSubquery` was changed to track the parent as well, so we could detect when we're analyzing a `SubqueryAlias` on the right side of a join, and add the left child to the scope. 

Additionally, pushdownFilters was modified to not undo all the changes to the Analyzer for HashLookups.

At runtime, the `PrependRow` nodes cache the rows outside the `InSubquery`, while the buildJoinIter for `CrossJoin` would include both the outside and the left row. This meant that some parts of the inner HashJoin would receive extra columns while others didn't. The fix here was to alter the `scope.InJoin` depending on which parts of HashJoin we were building.

Lastly, to have these changes not break for PreparedStatements, we just needed to not redo `finalizeUnions` in `postPrepared`, as we don't replan joins in `postPrepared`, so we don't know if we're in a join or not, and the correct indexes are set in `prePrepared`.

Along the way, we discovered a query that panics, but the cause is different than the purpose of this fix, and it panicked the same way before these changes, so it is left as a skipped test.

Fix for: https://github.com/dolthub/dolt/issues/5714

### CODE REVIEW

File: sql/analyzer/resolve_subqueries.go [hunk 1]
```diff
@@ -50,9 +50,31 @@ func finalizeSubqueries(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
...
+	}
+
+	var conFunc transform.CtxFunc = func(c transform.Context) (sql.Node, transform.TreeIdentity, error) {
+		n := c.Node
 		if sqa, ok := n.(*plan.SubqueryAlias); ok {
-			newSqa, same2, err := analyzeSubqueryAlias(ctx, a, sqa, scope, sel, true)
+			var newSqa sql.Node
+			var same2 transform.TreeIdentity
+			var err error
+			if sqa.OuterScopeVisibility && joinParent != nil && !joinParent.Op.IsLeftOuter() {
```
Comment (max-hoffman) [inline sql/analyzer/resolve_subqueries.go:67]: `!joinParent.Op.IsLeftOuter()` seems suspicious, why would any of the indexing be dependent on the parent being a left join? We should probably add a test for left joins.

File: sql/analyzer/resolve_subqueries.go [hunk 2]
```diff
@@ -50,9 +50,31 @@ func finalizeSubqueries(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
...
+				if stripChild, ok := joinParent.Right().(*plan.StripRowNode); ok && stripChild.Child == sqa {
+					subScope := scope.newScopeInJoin(joinParent.Children()[0])
```
Comment (max-hoffman) [inline sql/analyzer/resolve_subqueries.go:69]: We should add a comment that this only fixes 1-degree of join nesting, and that a proper fix would re-index the entire tree at the end of analysis. If the subquery has more than 1 sibling, or a sibling higher in the join tree, we currently do not detect it. We probably don't need tests for those edge cases yet. Just gotta hope no one hits them soon.

File: enginetest/queries/queries.go [hunk 3]
```diff
@@ -7522,6 +7522,37 @@ SELECT * FROM my_cte;`,
 			{2},
 		},
 	},
+	{
+		Query: `
```
Comment (max-hoffman) [inline enginetest/queries/queries.go:7526]: a comment tagging the original issue might be worthwhile

**Review by max-hoffman — COMMENTED** [review]
The PR summary is really good. The fixes are painful but maybe the only way to do this short of rewriting indexing. A comment pointing out how this fixes one layer of bugs from a larger general class of problems would be good. And then one more pass making sure we only target a specific edge case and are not causing collateral damage to currently working queries would be good. @zachmu this could use a pass, this is mostly my design and I hate it but I don't know a better way short of rewriting indexing.

File: sql/analyzer/exec_builder.go [hunk 4]
```diff
@@ -180,10 +180,18 @@ func (b *ExecBuilder) buildHashJoin(j *hashJoin, input sql.Schema, children ...s
 	if err != nil {
 		return nil, err
 	}
+	oldInJoin := false
+	if j.g.m.scope != nil {
```
Comment (zachmu) [inline sql/analyzer/exec_builder.go:184]: Rather than temporarily monkeying with the scope like this, why not pass a modified copy directly to this function?

File: sql/analyzer/resolve_subqueries.go [hunk 5]
```diff
@@ -420,10 +443,27 @@ func setJoinScopeLen(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
 	if scopeLen == 0 {
 		return n, transform.SameTree, nil
 	}
+
+	joinlessScopeLen := scopeLen
+	if scope.inJoin {
+		scope.SetJoin(false)
```
Comment (zachmu) [inline sql/analyzer/resolve_subqueries.go:449]: Same comment. Use these as values to compute things you need, rather than temporarily modifying

**Review by zachmu — COMMENTED** [review]
My one general comment is that while this is fine tactically, it seems like we are way overdue for general transform logic in the analyzer to consider sibling nodes correctly. Similar bugs are going to continue biting us until we fix this.

File: sql/analyzer/resolve_subqueries.go (hunk 1 above)
Comment (JCOR11599) [reply sql/analyzer/resolve_subqueries.go]: The `joinParent.Op.IsLeftOuter()` check might've been added to prevent breaking other queries in a previous iteration, but it looks like it isn't needed now. I'll remove it.

File: sql/analyzer/exec_builder.go [hunk 6]
```diff
@@ -180,7 +180,12 @@ func (b *ExecBuilder) buildHashJoin(j *hashJoin, input sql.Schema, children ...s
 	if err != nil {
 		return nil, err
 	}
-	outerAttrs, err := b.buildFilters(j.g.m.scope, j.right.relProps.OutputCols(), expression.Tuple(j.outerAttrs))
+	tmpScope := j.g.m.scope
+	if tmpScope != nil {
+		tmpScope = tmpScope.newScopeInJoin(nil)
```
Comment (max-hoffman) [inline sql/analyzer/exec_builder.go:185]: could we have slightly more expressive methods? like `j.g.m.scope.newScopeNoJoin()`?

**Review by max-hoffman — APPROVED** [review]
LGTM
